File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos02/Alg02.js
/*
♦ Algoritmo 02: Estrutura de Dados Linear Estática de Fila Circular e Não-Circular: Queue (Fila).

obs: Em uma fila circular é necessário administrar os elementos na fila em relação à exclusão de espaços 
    livres do início, para obter, de forma circular, mais espaços no fim, para adição de novos elementos.

obs: Para implementar uma fila não-circular, basta apenas deixar a fila encher sem utilizar os espaços 
    livres, caso haja, no início da fila.
*/

class FilaCircular {
    constructor(capacidade) {
        this.capacidade = capacidade;
        this.elementos = new Array(capacidade);
        this.tamanho = 0;
        this.frente = 0;
        this.traseira = -1;
    }

    // Verificar se a fila está vazia
    isEmpty() {
        return this.tamanho === 0;
    }

    // Verificar se a fila está cheia
    isFull() {
        return this.tamanho === this.capacidade;
    }

    // Adicionar elementos à fila (enqueue)
    enqueue(elemento) {
        if (this.isFull()) {
            throw new Error("A fila está cheia");
        }

        this.traseira = (this.traseira + 1) % this.capacidade;
        this.elementos[this.traseira] = elemento;
        this.tamanho++;
    }

    // Remover o elemento da frente da fila (dequeue)
    dequeue() {
        if (this.isEmpty()) {
            throw new Error("A fila está vazia");
        }

        const elementoRemovido = this.elementos[this.frente];
        this.elementos[this.frente] = null;
        this.frente = (this.frente + 1) % this.capacidade;
        this.tamanho--;

        return elementoRemovido;
    }

    // Obter o tamanho atual da fila
    size() {
        return this.tamanho;
    }

    // Acessar o elemento da frente da fila sem removê-lo
    front() {
        if (this.isEmpty()) {
            throw new Error("A fila está vazia");
        }

        return this.elementos[this.frente];
    }

    // Mostrar elementos da fila
    showQueue() {
        let index = this.frente;
        for (let i = 0; i < this.tamanho; i++) {
            console.log(this.elementos[index] + " ");
            index = (index + 1) % this.capacidade;
        }
        console.log();
    }
}

const fila = new FilaCircular(5);
fila.enqueue(1);
fila.enqueue(2);
fila.enqueue(3);

fila.showQueue();

console.log("Frente da fila: " + fila.front());
console.log("Tamanho da fila: " + fila.size());

fila.dequeue();
fila.showQueue();

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos02/Alg02.rb
=begin 
♦ Algoritmo 02: Estrutura de Dados Linear Estática de Fila Circular e Não-Circular: Queue (Fila).

obs: Em uma fila circular é necessário administrar os elementos na fila em relação à exclusão de espaços 
    livres do início, para obter, de forma circular, mais espaços no fim, para adição de novos elementos.

obs: Para implementar uma fila não-circular, basta apenas deixar a fila encher sem utilizar os espaços 
    livres, caso haja, no início da fila.
=end

class FilaCircular
  def initialize(capacidade)
    @capacidade = capacidade
    @elementos = Array.new(capacidade)
    @tamanho = 0
    @frente = 0
    @traseira = -1
  end

  # Verificar se a fila está vazia
  def is_empty?
    @tamanho == 0
  end

  # Verificar se a fila está cheia
  def is_full?
    @tamanho == @capacidade
  end

  # Adicionar elementos à fila (enqueue)
  def enqueue(elemento)
    if is_full?
      raise "A fila está cheia"
    end

    @traseira = (@traseira + 1) % @capacidade
    @elementos[@traseira] = elemento
    @tamanho += 1
  end

  # Remover o elemento da frente da fila (dequeue)
  def dequeue
    if is_empty?
      raise "A fila está vazia"
    end

    elemento_removido = @elementos[@frente]
    @elementos[@frente] = nil
    @frente = (@frente + 1) % @capacidade
    @tamanho -= 1

    elemento_removido
  end

  # Obter o tamanho atual da fila
  def size
    @tamanho
  end

  # Acessar o elemento da frente da fila sem removê-lo
  def front
    if is_empty?
      raise "A fila está vazia"
    end

    @elementos[@frente]
  end

  # Mostrar elementos da fila
  def show_queue
    index = @frente
    @tamanho.times do
      print "#{@elementos[index]} "
      index = (index + 1) % @capacidade
    end
    puts
  end
end

# Exemplo de uso
fila = FilaCircular.new(5)
fila.enqueue(1)
fila.enqueue(2)
fila.enqueue(3)

fila.show_queue

puts "Frente da fila: #{fila.front}"
puts "Tamanho da fila: #{fila.size}"

fila.dequeue
fila.show_queue
